Netcore with dotnetcli

enricosada opened the pull request:
from: https://plus.google.com/events/c6tik5o4aofom0s2numam7nc57c

I updated nunit to version 3.
nunit projects now are executable in project.json (fsproj are not changed) with nunit runner, you just need to run it (xplat works!)

the dotnet build expect the `paket-files/test/forki/FsUnit/FsUnit.fs` already downloaded

enricosada committed:
dotnet new

enricosada committed:
add source files

enricosada committed:
it's a library

enricosada committed:
move deps under frameworks.dnxcore50

enricosada committed:
ignore project.lock.json

enricosada committed:
add .NET 4.6

enricosada committed:
add project.json tests

enricosada committed:
the error FS0193: internal error: Exception of type 'System.Collections.Generic.KeyNotFoundException' was thrown mean missing FSharp.Core.dll

enricosada committed:
add Chessie as project ref. but restore doesnt work because .net cli search sibling directory

enricosada committed:
add a global.json and run dotnet restore from root

enricosada committed:
add deps to dnxcore50

enricosada committed:
refactor dependencies. the global dependencies are inherited by frameworks

enricosada committed:
nunit 3 use nunitlite to run test suite, the test project are executables

enricosada committed:
convert Chessie.CSharp.Test

enricosada committed:
set package metadata

enricosada committed:
add appveyor build

enricosada committed:
add travis build

mexx commented on .travis.yml:
```diff
@@ -1,6 +1,51 @@
   - ./build.sh All
+  # Build .NET Core with .NET CLI
+  # Workaround fsc doesnt work
+  - if test "$TRAVIS_OS_NAME" == "osx"; then sudo curl -L -s https://github.com/enricosada/fsharp-dotnet-cli-samples/raw/master/workarounds/1/osx.10.10-x64.zip | tar xvf - -C /usr/local/share/dotnet/cli/bin/ ; fi
```
Shouldn't this go into the `before_install` section?

enricosada replied:
oki

mexx commented on .travis.yml:
```diff
@@ -1,6 +1,51 @@
+  - ulimit -n 1024
+  # cli 
+  - dotnet restore
+  # Build Chessie
```
Can you put `Build Chessie` and following test runs into the build script itself, so it can also be run locally the same way it does on the CI server?
Same thing applies to the changes of `appveyor.yml` .

enricosada replied:
If possibile, i dont want to change your build script.
If it's needed i'll do it obv, but as another pr, it's easier to review

There are two parts:

### install .NET CLI
- the user may have it installed locally, no need to install always
- you can add an install script in your `scripts` directory, but it's annoying to maintain and it's two lines (the install script is downloaded from dotnet cli repo)
- if you want do that, use a `.dotnet` subdirectory instead of publish to default installation directory (i can help with that if you want, it's an argument of install script).
- usually the global .net cli install is enough, for shared ci is better a `.netcli` directory but appveyor/travis are always new images, so it's not important. But maybe it's better a `.netcli` subdir because it's easier to redo steps locally without change current global installation

### build and run with .NET CLI
- if possibile i dont want to touch your build script because i dont know enough of it. 
- i think the best is to add a fake task to run the same commands (not dotnet cli install, i mean `dotnet restore` and `dotnet build` ) /cc @forki for suggestions
- the current FAKE package creation task should include the binaries (you can use `dotnet build --output` to change directory for binaries if you want put it near current .NET 4.6 binaries)
- the new FAKE task should run conditionally i think because not everyone has .NET cli installed

mexx replied:
Install of .NET CLI should be environment specific, it's ok to have it in the CI server file.

About build and test with .NET CLI, I would put this part into the build script, and also as part of this PR. Why do we need an other PR for moving things in the right places, when it's already clear now?

> if possibile i dont want to touch your build script because i dont know enough of it.

Provide two targets in the build script, one for Build and one for Test. There is nothing special about the build script, it's a standard FAKE script from project scaffold.

forki commented on src/Chessie/project.json:
```diff
@@ -0,0 +1,27 @@
+    ],
+
+    "frameworks": {
+        "net46": {
```
why do we have net46 at all?

enricosada replied:
I use net46 as check for project.json before migrate libraries to .NET Core
If net46 works, dotnet cli works. 
After that, dnxcore errors are about migration full .net -> netcore ( or fsc bugs, but not that common )

Just pass `--framework dnxcore50` to `dotnet build`

enricosada committed:
add .net cli build inside fake script

enricosada committed:
remove .net cli commands from travis/appveyor, now are inside the build script

enricosada commented:
added `DotnetCliBuild` and `DotnetCliRunTests` targets
these target run if dotnet cli is installed (check with `dotnet --version`)

I dont know how to modify the nuget package, because it's generated by `paket.templates` file.

if you run:

``` bash
dotnet restore
cd src/Chessie
dotnet pack --configuration Release
```

you can see the nuget package generated with the right nuspec (with both net46 and dnxcore50)

enricosada committed:
pin version of install script

enricosada committed:
add assemblies to nupkg

enricosada commented:
Ok now package contains the dnxcore assemblies too

But i dont know how to specify dependencies by framework /cc @forki for help

i tried in `paket.dependencies`

```
nuget FSharp.Core redirects:force, framework: net45
nuget Microsoft.FSharp.Core.netcore = 1.0.0-alpha-151221 framework: dnxcore50
nuget NETStandard.Library = 1.0.0-rc2-23811 framework: dnxcore50
```

but nuspec is the same

forki commented:
That is not supported yet.
On Mar 22, 2016 8:10 PM, "Enrico Sada" notifications@github.com wrote:

> Ok now package contains the dnxcore assemblies too
> 
> But i dont know how to specify dependencies by framework /cc @forki
> https://github.com/forki for help
> 
> i tried in paket.dependencies
> 
> nuget FSharp.Core redirects:force, framework: net45
> nuget Microsoft.FSharp.Core.netcore = 1.0.0-alpha-151221 framework: dnxcore50
> nuget NETStandard.Library = 1.0.0-rc2-23811 framework: dnxcore50
> 
> but nuspec is the same
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/fsprojects/Chessie/pull/43#issuecomment-199968843

enricosada commented:
thx @forki, i'll add a workaround so, good old search/replace

enricosada committed:
workaround add netcore to nuget package

enricosada commented:
ok added a workaround, the target `AddNetcoreToNupkg` update the nupkg generated from paket adding `dnxcore50` framework, files and dependancies  /cc @mexx @forki

the deps use `group` element

``` xml
    <dependencies>
      <group targetFramework=".NETFramework4.0">
        <dependency id="FSharp.Core" version="" />
      </group>
      <group targetFramework="DNXCore5.0">
        <dependency id="Microsoft.FSharp.Core.netcore" version="[1.0.0-alpha-151221, )" />
        <dependency id="NETStandard.Library" version="[1.0.0-rc2-23811, )" />
      </group>
    </dependencies>
```

enricosada commented:
wait before merge, i want to extract the hack about nuspec as gist (thx @forki for the idea)

enricosada committed:
use dotnet-mergenupkg tool to merge nupkgs

enricosada committed:
use netstandard1.5 instead of dnxcore50

enricosada commented:
Ok a bit late :smile: 
Changed fw from `dnxcore50` to `netstandard1.5` (that's the latest fashion)
.net cli version: `1.0.0-beta-002071` 

removed the custom code with `dotnet-mergenupkg` to add .net cli deps (assemblies + nuspec deps) to the already built .nupkg

enricosada committed:
update travis/appveyor

enricosada commented:
@mexx i need to do something? rebase/etc?

mexx commented:
@enricosada I'm actually not involved in the development, redirecting your question to @forki.

forki commented:
@enricosada think it's ready to merge and release?

sergey-tihon commented:
Would be awesome to have such integration on Travis CI build matrix level and be able to write something like this:

``` yaml
language: csharp

os:
  - linux
  - osx

mono:
  - latest
  - 4.2.2

dotnet-core:
  - latest
  - 1.0.0-beta-002071
```

Does it make sense? Could we request\implement it on Travis?
